src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.Worker_ResolveReferences.cs

```diff
@@ -0,0 +1,291 @@
+                            // Leave the metadata reference and don't create a project reference.
+                            var outputFilePath = await _buildManager.TryGetOutputFilePathAsync(absoluteProjectPath, _globalProperties, cancellationToken).ConfigureAwait(false);
+                            if (!string.IsNullOrEmpty(outputFilePath) &&
+                                metadataReferenceSet.Contains(outputFilePath) &&
+                                File.Exists(outputFilePath))
+                            {
+                                continue;
+                            }
+                        }
+
+                        // If metadata is preferred or there's no loader for this project, see if the output path exists.
+                        // If it does, don't create a project reference.
+                        if (_preferMetadataForReferencedProjects)
+                        {
+                            var projectRefFileInfos = await LoadProjectFileInfosAsync(
+                                absoluteProjectPath, DiagnosticReportingOptions.IgnoreAll, cancellationToken).ConfigureAwait(false);
+
+                            var done = false;
+                            foreach (var projectRefFileInfo in projectRefFileInfos)
+                            {
+                                if (!string.IsNullOrEmpty(projectRefFileInfo.OutputFilePath) && metadataReferenceSet.Contains(projectRefFileInfo.OutputFilePath))
+                                {
+                                    // We found it!
+                                    if (File.Exists(projectRefFileInfo.OutputFilePath))
+                                    {
+                                        done = true;
+                                        break;
+                                    }
+                                }
+                            }
+
+                            if (done)
+                            {
+                                continue;
+                            }
+
+                            // We didn't find the output file path in this project's metadata references, or it doesn't exist on disk.
+                            // In that case, carry on and load the project.
+                        }
+
+                        // OK, we've got to load the project.
+                        var projectRefInfos = await LoadProjectInfosFromPathAsync(absoluteProjectPath, _discoveredProjectOptions, cancellationToken).ConfigureAwait(false);
+
+                        var projectRefInfo = metadataReferenceSet.Find(projectRefInfos);
+                        if (projectRefInfo != null)
+                        {
+                            // Don't add a reference if the project already has a reference on us. Otherwise, it will cause a circularity.
+                            if (projectRefInfo.ProjectReferences.Any(pr => pr.ProjectId == id))
```

Comment:
Not looking for transitive circularities?

Reply:
MSBuildWorkspace didn't do that in the past, so I didn't bother to add it. The original logic was, if the projects reference one another, one of them needs to become a metadata reference. Do you feel strongly about this?

Reply:
Fine with me.